web/source/keyboards/keyboard.ts

```diff
@@ -241,5 +262,69 @@ namespace com.keyman.keyboards {
         this.scriptObject['KNS'](_PCommand, _PTarget, _PData);
       }
     }
+
+    private findOrConstructLayout(formFactor: text.FormFactor): LayoutFormFactor {
+      if(this._layouts) {
+        if(this._layouts[formFactor]) {
+          return this._layouts[formFactor];
+        } else if(formFactor == text.FormFactor.Phone && this._layouts[text.FormFactor.Tablet]) {
+          return this._layouts[text.FormFactor.Phone] = this._layouts[text.FormFactor.Tablet];
+        } else if(formFactor == text.FormFactor.Tablet && this._layouts[text.FormFactor.Phone]) {
+          return this._layouts[text.FormFactor.Tablet] = this._layouts[text.FormFactor.Phone];
+        }
+      }
+
+      // No pre-built layout available; time to start constructing it via defaults.
+      // First, if we have non-default keys specified by the ['BK'] array, we've got
+      // enough to work with to build a default layout.
+      let rawSpecifications: any = null;  // TODO:  better typing, same type as this.legacyLayoutSpec.
+      if(this._legacyLayoutSpec != null && this._legacyLayoutSpec['BK'] != null) {
+        var keyCaps=this._legacyLayoutSpec['BK'];
+        for(var i=0; i<keyCaps.length; i++) {
+          if(keyCaps[i].length > 0) {
+            rawSpecifications = this._legacyLayoutSpec;
+            break;
+          }
+        }
+      }
+
+      // If we don't have key definitions to use for a layout but also lack help text or are a touch-based layout,
+      // we make a default layout anyway.  We have to show display something usable.
+      if(!rawSpecifications && (this.helpText == '' || formFactor != text.FormFactor.Desktop)) {
+        rawSpecifications = {'F':'Tahoma', 'BK': Layouts.dfltText};
+      }
+
+      // Final check - do we construct a layout, or is this a case where helpText / insertHelpHTML should take over?
+      if(rawSpecifications) {
+        if(!this._layouts) {
+          // Initialize the field that backs the property; may as well cache the default layout we just built.
+          this._layouts = {};
+        }
+        // Now to generate a layout from our raw specifications.
+        let layout = this._layouts[formFactor] = Layouts.buildDefaultLayout(rawSpecifications, this, formFactor);
+        layout.isDefault = true;
+        return layout;
+      } else {
+        // The fact that it doesn't exist will indicate that help text/HTML should be inserted instead.
+        return null;
+      }
+    }
+
+    /**
+     * Returns an ActiveLayout object representing the keyboard's layout for this form factor.  May return null if a custom desktop "help" OSK is defined, as with sil_euro_latin.
+     * 
+     * In such cases, please use either `helpText` or `insertHelpHTML` instead.
+     * @param formFactor {string} The desired form factor for the layout.
+     */
+    public layout(formFactor: text.FormFactor): ActiveLayout {
+      let rawLayout = this.findOrConstructLayout(formFactor);
+
+      if(rawLayout) {
```

Comment:
An argument could be made for more overtly 'caching' this; the `polyfill` call does do some computation, so settings flags for layouts that have already been processed could be used to prevent overhead should this method get overused.

Reply:
... actually, I should do that, as it will reset the small adjustments made to fit the OSK nicely to the form factor.